src/modules/navigator/precland.cpp

```diff
@@ -0,0 +1,580 @@
+		(void)orb_unadvertise(h);
+
+		_state = PrecLandState::FinalApproach;
+		_state_start_time = hrt_absolute_time();
+		return true;
+	}
+	return false;
+}
+
+bool
+PrecLand::switch_to_state_search()
+{
+	PX4_INFO("Climbing to search altitude.");
+	vehicle_local_position_s * vehicle_local_position = _navigator->get_local_position();
+
+	position_setpoint_triplet_s *pos_sp_triplet = _navigator->get_position_setpoint_triplet();
+	pos_sp_triplet->current = _first_sp;
+	pos_sp_triplet->current.alt = vehicle_local_position->ref_alt + _param_search_alt.get();
+	pos_sp_triplet->current.type = position_setpoint_s::SETPOINT_TYPE_POSITION;
+	_navigator->set_position_setpoint_triplet_updated();
+
+	_beacon_acquired_time = 0;
+
+	_state = PrecLandState::Search;
+	_state_start_time = hrt_absolute_time();
+	return true;
+}
+
+bool
+PrecLand::switch_to_state_fallback()
+{
+	PX4_WARN("Falling back to normal land.");
+	position_setpoint_triplet_s *pos_sp_triplet = _navigator->get_position_setpoint_triplet();
+	pos_sp_triplet->current = _first_sp; // should this be at current position?
+	pos_sp_triplet->current.type = position_setpoint_s::SETPOINT_TYPE_LAND;
+	_navigator->set_position_setpoint_triplet_updated();
+
+	_state = PrecLandState::Fallback;
+	_state_start_time = hrt_absolute_time();
+	return true;
+}
+
+bool PrecLand::check_state_conditions(PrecLandState state)
+{
+	vehicle_local_position_s * vehicle_local_position = _navigator->get_local_position();
+
+	switch(state) {
+		case PrecLandState::Start:
+			return _search_cnt <= _param_max_searches.get();
+
+		case PrecLandState::HorizontalApproach:
+			return _beacon_position_valid && _beacon_position.abs_pos_valid;
+		
+		case PrecLandState::DescendAboveBeacon:
+			// if we're already in this state, only leave it if beacon becomes unusable, don't care about horizontall offset to beacon
+			if (_state == PrecLandState::DescendAboveBeacon)
+			{
+				// if we're close to the ground, we're more critical of beacon timeouts so we quickly go into descend
+				if (check_state_conditions(PrecLandState::FinalApproach))
+				{
+					return hrt_absolute_time() - _beacon_position.timestamp < 500000; // 0.5s
+				} else {
+					return _beacon_position_valid && _beacon_position.abs_pos_valid;
+				}
+			} else {
+				// if not already in this state, need to be above beacon to enter it
+				return _beacon_position_valid && _beacon_position.abs_pos_valid
+				&& fabsf(_beacon_position.x_rel) < _param_hacc_rad.get() && fabsf(_beacon_position.x_rel) < _param_hacc_rad.get();
+			}
+		
+		case PrecLandState::FinalApproach:
+			return vehicle_local_position->dist_bottom_valid && vehicle_local_position->dist_bottom < _param_final_approach_alt.get();
+		
+		case PrecLandState::Search:
+			return true;
+		
+		case PrecLandState::Fallback:
+			return true;
+		
+		default:
+			return false;
+	}
+}
+
+void PrecLand::slewrate(float &sp_x, float &sp_y)
```

Comment:
couldn't we replaces this with `math::gradual`?

Reply:
this does more than `math::gradual`. It does not simply interpolate linearly but rather quadratically. It respects velocity and acceleration limits and reduces setpoint velocity in anticipation of reaching the target setpoint s.t. acceleration limits are respected.

I was thinking about adding this in an `else` to [this](https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L1982) or enabling [this](https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L2475) for auto modes. That would have effects on other flight modes and would require extensive testing.